File: 01-Languages/Cpp/22-Mini-Projects/TaskTracker/main.cpp
// main.cpp - the CLI entry point: hand-rolled argv parsing (C++ has no built-in CLI
// argument-parsing library, an honest gap in the same spirit as Lessons 10/16/17's
// "no built-in JSON/database/HTTP client" observations), dispatching to a thin
// TaskRepository call per command. Deliberately kept free of any SQL/business logic
// itself, so TaskRepository stays independently testable (see tests/test_TaskRepository.cpp).
#include <iostream>
#include <iomanip>
#include <vector>
#include <string>
#include <sstream>
#include "TaskRepository.hpp"
#include "TaskNotFoundException.hpp"

void printUsage() {
    std::cout <<
        "Usage:\n"
        "  app add <title> [--priority low|medium|high]\n"
        "  app list [--status pending|done]\n"
        "  app done <id>\n"
        "  app delete <id>\n"
        "  app stats\n";
}

void printTask(const TaskItem& t) {
    std::cout << (t.status == Status::Done ? "[x] " : "[ ] ")
               << "#" << t.id << "  "
               << std::left << std::setw(28) << t.title
               << "priority=" << std::setw(7) << toString(t.priority)
               << "created=" << t.createdAt << std::endl;
}

int main(int argc, char* argv[]) {
    std::vector<std::string> args(argv + 1, argv + argc);
    if (args.empty()) {
        printUsage();
        return 0;
    }

    TaskRepository repo("tasks.db");
    const std::string& command = args[0];

    try {
        if (command == "add") {
            if (args.size() < 2) { std::cout << "Error: 'add' requires a title.\n"; return 1; }
            std::string title = args[1];
            Priority priority = Priority::Medium;
            for (size_t i = 2; i + 1 < args.size(); ++i) {
                if (args[i] == "--priority") priority = parsePriority(args[i + 1]);
            }
            int id = repo.addTask(title, priority);
            std::cout << "Added task #" << id << ": " << title << " (priority=" << toString(priority) << ")\n";

        } else if (command == "list") {
            std::optional<Status> filter;
            for (size_t i = 1; i + 1 < args.size(); ++i) {
                if (args[i] == "--status") filter = parseStatus(args[i + 1]);
            }
            auto tasks = repo.listTasks(filter);
            if (tasks.empty()) {
                std::cout << "(no tasks)\n";
            } else {
                for (const auto& t : tasks) printTask(t);
            }

        } else if (command == "done") {
            if (args.size() < 2) { std::cout << "Error: 'done' requires an id.\n"; return 1; }
            int id = std::stoi(args[1]);
            repo.markDone(id);
            std::cout << "Marked task #" << id << " as done.\n";

        } else if (command == "delete") {
            if (args.size() < 2) { std::cout << "Error: 'delete' requires an id.\n"; return 1; }
            int id = std::stoi(args[1]);
            repo.deleteTask(id);
            std::cout << "Deleted task #" << id << ".\n";

        } else if (command == "stats") {
            TaskStats stats = repo.getStats();
            std::cout << "Pending: " << stats.pending << "  Done: " << stats.done
                       << "  Total: " << stats.total() << "\n";

        } else {
            printUsage();
        }
    } catch (const TaskNotFoundException& e) {
        // Caught separately from the generic branch below so this specific, expected
        // failure mode (a bad id typed by the user) reads as "Error: ..." rather than
        // an unhandled-crash-looking message -- exactly the polymorphic-catch pattern
        // from 20-Exercises/Exercise 04, applied to a real CLI instead of a toy demo.
        std::cout << "Error: " << e.what() << "\n";
        return 1;
    } catch (const std::exception& e) {
        std::cout << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: 01-Languages/Kotlin/22-Mini-Projects/TaskTracker/src/Main.kt
import java.sql.DriverManager

private const val USAGE = """Usage:
  add <title> [--priority low|medium|high]
  list [--status pending|done]
  done <id>
  delete <id>
  stats"""

fun main(args: Array<String>) {
    if (args.isEmpty()) {
        println(USAGE)
        return
    }

    Class.forName("org.sqlite.JDBC") // registers the SQLite JDBC driver -- same requirement as Lesson 16
    DriverManager.getConnection("jdbc:sqlite:tasks.db").use { conn ->
        val repo = TaskRepository(conn)
        try {
            dispatch(repo, args)
        } catch (e: TaskNotFoundException) {
            println("Error: ${e.message}")
        } catch (e: IllegalArgumentException) {
            println("Error: ${e.message}")
        }
    }
}

private fun dispatch(repo: TaskRepository, args: Array<String>) {
    when (args[0]) {
        "add" -> {
            val title = args.getOrNull(1) ?: run { println("Error: add requires a title"); return }
            val priority = flagValue(args, "--priority")?.let { parsePriority(it) } ?: Priority.MEDIUM
            val task = repo.addTask(title, priority)
            println("Added task #${task.id}: ${task.title} (priority=${task.priority.display()})")
        }
        "list" -> {
            val status = flagValue(args, "--status")?.let { parseStatus(it) }
            val tasks = repo.listTasks(status)
            if (tasks.isEmpty()) {
                println("No tasks.")
            } else {
                for (task in tasks) println(task.format())
            }
        }
        "done" -> {
            val id = requireIdArg(args)
            repo.markDone(id)
            println("Marked task #$id as done.")
        }
        "delete" -> {
            val id = requireIdArg(args)
            repo.deleteTask(id)
            println("Deleted task #$id.")
        }
        "stats" -> {
            val stats = repo.getStats()
            println("Pending: ${stats.pending}  Done: ${stats.done}  Total: ${stats.total}")
        }
        else -> println(USAGE)
    }
}

private fun requireIdArg(args: Array<String>): Long =
    args.getOrNull(1)?.toLongOrNull() ?: throw IllegalArgumentException("expected a numeric task id")

// A tiny, hand-rolled --flag value lookup -- proportionate to this app's five commands and two
// flags; a real CLI with more surface area would reach for a library instead of growing this by hand.
private fun flagValue(args: Array<String>, flag: String): String? {
    val index = args.indexOf(flag)
    return if (index >= 0 && index + 1 < args.size) args[index + 1] else null
}

private fun parsePriority(raw: String): Priority = when (raw.lowercase()) {
    "low" -> Priority.LOW
    "medium" -> Priority.MEDIUM
    "high" -> Priority.HIGH
    else -> throw IllegalArgumentException("unknown priority '$raw' (expected low|medium|high)")
}

private fun parseStatus(raw: String): Status = when (raw.lowercase()) {
    "pending" -> Status.PENDING
    "done" -> Status.DONE
    else -> throw IllegalArgumentException("unknown status '$raw' (expected pending|done)")
}

private fun Priority.display(): String = name.lowercase().replaceFirstChar { it.uppercase() }

private fun TaskItem.format(): String {
    val mark = if (status == Status.DONE) "x" else " "
    return "[$mark] #$id\t$title\tpriority=${priority.display()}\tcreated=${createdAt.take(10)}"
}
